### PR TITLE
Create dicom orient image filter

### PR DESCRIPTION
**Pull request by blowekamp**
The is a refactoring of the itk::OrientImageFilter. This changes the meaning of the orientation directions so that the meaning matches that of DICOM. An ITK image with an identity direction cosine matrix will be in LPS coordinates.

This is work to address  SimpleITK/SimpleITK#644 and InsightSoftwareConsortium/ITK#1163.

### CODE REVIEW

**Comment by blowekamp** [comment]
@hjmjohnson I believe you worked on the original itk::OrientImageFilter. I hope this refactoring can overcome some of the discrepancies/limitation in the orginal and be a filter which works well with the current ITK orientation information to provide an easy way to approximate re-orientation of images.

Feedback on this work is welcomed.

Clearly, extensive testing is still needed.

**Comment by gdevenyi** [comment]
Is there a reason this is a PR against SimpleITK instead of ITK itself?

**Comment by blowekamp** [comment]
New filters generally are suppose to go into remote modules first. Let's try to get it into this remote module for ITKv5.1. Then we can bring up the issue if it should be added to ITK for ITKv5.2 after the filter as been used and "stabilized."

**Comment by gdevenyi** [comment]
> New filters generally are suppose to go into remote modules first.

@gdevenyi read the ITK developer's guide :) Got it :+1: I should do that at some point for other reasons.

**Comment by blowekamp** [comment]
This filter should be just about ready to go. @zivy @gdevenyi  Please review.

**Comment by gdevenyi** [comment]
@blowekamp is there a quickish way to build/test this in python? I don't think I have the build system to build python packages

**Comment by gdevenyi** [comment]
@Gab-D-G this should be able to replace nibabel's orientation filter once merged.

File: test/itkDICOMOrientImageFilterGTest.cxx [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+  using ImageType = itk::Image<uint32_t, 3>;
+
+  auto sourceFilter = itk::RandomImageSource<ImageType>::New();
+  sourceFilter->SetMin(-1000);
+  sourceFilter->SetMax(1000);
+  sourceFilter->SetSize({ 5, 6, 7 });
+
+
+  using OrientFilterType = itk::DICOMOrientImageFilter<ImageType>;
+  using PermuiteArrayType = OrientFilterType::PermuteOrderArrayType;
```
Comment (zivy) [inline test/itkDICOMOrientImageFilterGTest.cxx]: remove the i (PermuiteArrayType -> PermuteArrayType), but can be ignored as the error is consistent in the code, so no real harm.

File: include/itkDICOMOrientImageFilter.h [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+  enum class CoordinateEnum : uint8_t
+  {
+    UNKNOWN = 0,
+    Right = 2, // 0b0010
+    Left = 3,
+    Posterior = 4, // back - 0b0100
+    Anterior = 5,  // front
+    Inferior = 8,  // below - 0b1000
+    Superior = 9   // above
+  };
```
[the remark this answers is not in the document]
Comment (blowekamp) [reply include/itkDICOMOrientImageFilter.h]: done

File: include/itkDICOMOrientImageFilter.h [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+  {
+    auto            iter = m_StringToCode.find(desired);
+    OrientationEnum desiredEnum = OrientationEnum::INVALID;
+    if (iter != m_StringToCode.end())
+    {
+      desiredEnum = iter->second;
+    }
+    else
+    {
+      itkWarningMacro("Invalid desired coordinate direction string: \"" << desired << "\"!");
```
[the remark this answers is not in the document]
Comment (blowekamp) [reply include/itkDICOMOrientImageFilter.h]: If it's invalid then the filter with just not perform any flip or permutations. We can refine this behavior later.

File: test/itkDICOMOrientImageFilterGTest.cxx (hunk 1 above)
Comment (blowekamp) [reply test/itkDICOMOrientImageFilterGTest.cxx]: If it's invalid then the filter with just not perform any flip or permutations. We can refine this behavior later.

**Pull request merged**